### DISCUSSION TITLE
Raccolta differenziata

### DISCUSSION THREAD

**Discussion by giros79** [indomus/forum#1305] [HUB: Home Assistant]
Buongiorno a tutti, 
ho trovato in rete quanto mi serviva per gestire un sensore per la differenziata. Funziona tutto benissimo e sono soddisfatto.
Ho un solo problema, quando c'è un'alternanza (nel mio caso il martedì si alternano vetro e indifferenziato) quale è la sintassi giusta? 
PS: capisco poco e niente di sintassi

```
sensor:
  - platform: template
    sensors:
      rifiuti:
        friendly_name: "Garbage"
        icon_template: mdi:recycle
        value_template: "
        {% if (now().timestamp() | int - (60)) | timestamp_custom('%w') == '1' %} Plastica
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '2' %} Indifferenziato?????????????????????????
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '3' %} Organico
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '4' %} Carta
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '5' %} Organico
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '6' %} Niente
        {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '0' %} Organico
        {% endif %}
        "
        entity_id: sensor.date_time
```

Grazie in anticipo a chi vorrà darmi una mano

**Comment by snakuzzo** [comment]
Per cortesia...correggi il post ed usa il [markdown](https://github.com/indomus/forum/blob/main/README.md) per postare il codice

**Reply by giros79** [reply]
thanks

**Reply by snakuzzo** [reply]
Se l'alternanza è così precisa, potresti verificare se il numero della settimana è pari o dispari con...

```python
now().isocalendar()[1] | int % 2
```

Quindi il sensore andrebbe modificato così (scambia tu le stringhe Vetro e Indifferenziato in base all'alternanza ovviamente)...


```yaml
sensor:
  - platform: template
    sensors:
      rifiuti:
        friendly_name: "Garbage"
        icon_template: mdi:recycle
        value_template: >
          {% if (now().timestamp() | int - (60)) | timestamp_custom('%w') == '1' %} Plastica
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '2' %}  a
            {% if now().isocalendar()[1] | int % 2 == 0 %}
              Vetro
            {% else%}
              Indifferenziato
            {% endif %}
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '3' %} Organico
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '4' %} Carta
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '5' %} Organico
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '6' %} Niente
          {% elif (now().timestamp() | int - (60)) | timestamp_custom('%w') == '0' %} Organico
          {% endif %}
```

Non so a che versione di HA sei, ma ho tolto anche...

```yaml
entity_id: sensor.date_time
```

visto che dovrebbe essere deprecato da parecchie versioni ormai. Tra le altre cose l'entity_id del tuo sensore sarà `sensor.rifiuti`

**Reply by giros79** [reply]
grazie mille!!!!!! e il bello è che ho anche capito. Il problema è che mi mancano i fondamentali :(

**Reply by snakuzzo** [reply]
un po' alla volta...se hai pazienza e dedizione vedrai che non sarà un problema.
Sono veramente concetti molto basilari...e poi ricorda..."Google è tuo amico" :D

**Reply by xyzmauro** [reply]
Posso dire “che brutto” veder usare now() per un sensore che cambia 1 volta al giorno? Così viene valutato ogni minuto e ha poco senso,forse meglio usare sensor.date che cambia a 1/2notte e per tutto il giorno resta così…🤷‍♂️

**Reply by snakuzzo** [reply]
Sono d'accordo. Tra l'altro è pure sconsigliato in documentazione se non ricordo male

**Comment by Giancky79** [comment]
io l'avrei fatta cosi

```
template:
  - sensor:
      - name: "Rifiuti"
        unique_id: "garbage"
        state: >
            {% set pari = ['Plastica', 'Vetro', 'Organico', 'Carta', 'Organico', 'Niente', 'Organico'] %} 
            {% set ritiro_dispari = 'Indifferenziato' %} 
            {% set ritiro_pari = pari[now().weekday()] %}
            {% if (((now().strftime("%W"))|int)%2 == 0) %}
              {{ ritiro_pari }}
            {% else %}
              {{ ritiro_dispari }}
            {% endif %}
```

**Reply by xyzmauro** [reply]
Sicuro Giancky? Così nella settimana dispari il sensore sarà sempre su “indifferenziata” indipendentemente dal giorno, invece l’esigenza è che cambi solo il rifiuto del Martedì a settimane alterne ma il resto deve restare uguale…🤔
PS la “critica” all’uso di now() resta valida 😅

**Reply by Giancky79** [reply]
Vero, fatta di fretta, mq non esiste altro modo per avere un sensore settimana, da come ho interpretato la documentazione è l’uso di now() o utcnow(), non con day o weekday etc, perché avresti uno stato aggiornato continuo col cambio del secondo, invece come con weekday lo stato è quello per 23:59:59, quindi non crea problemi, altrimenti che altre soluzioni ci sarebbero?

**Reply by Giancky79** [reply]
```
            {% set pari = ['Plastica', 'Vetro', 'Organico', 'Carta', 'Organico', 'Niente', 'Organico'] %} 
            {% set dispari = ['Plastica', 'Indifferenziato', 'Organico', 'Carta', 'Organico', 'Niente', 'Organico'] %} 
            {% set ritiro_pari = pari[now().weekday()] %}
            {% set ritiro_dispari = dispari[now().weekday()] %}
            {% if (((now().strftime("%W"))|int)%2 == 0)  %}
              {{ ritiro_pari }}
            {% else %}
              {{ ritiro_dispari }}
            {% endif %}
```

allora cosi, ma non ho altro per la soluzione now().weekday()

**Reply by xyzmauro** [reply]
> Vero, fatta di fretta, mq non esiste altro modo per avere un sensore settimana, da come ho interpretato la documentazione è l’uso di now() o utcnow(), non con day o weekday etc, perché avresti uno stato aggiornato continuo col cambio del secondo, invece come con weekday lo stato è quello per 23:59:59, quindi non crea problemi, altrimenti che altre soluzioni ci sarebbero?

Da come ho capito io invece anche con {{ now().weekday() }} ogni minuto viene ricalcolato il giorno della settimana e di conseguenza il valore del template, ed è inutile farlo quando cambia solo al passaggio del giorno...
Peggio ancora con {{ (now().strftime("%W")|int) }} con cui calcoli il numero della settimana ogni minuto e cambia solo una volta a settimana
{{ now().weekday() }} si può scrivere anche {{ as_timestamp(states('sensor.date')) | timestamp_custom('%w') | int -1}} e
{{ (now().strftime("%W")|int) }} così {{ as_timestamp(states('sensor.date')) | timestamp_custom('%W') | int }} usando il sensor.date che cambia una volta sola al giorno

**Reply by xyzmauro** [reply]
![Senza titolo](https://user-images.githubusercontent.com/56839356/150617596-43a30ba6-bf4a-427c-805c-ee78a6b93f1d.png)

**Reply by Giancky79** [reply]
Cercato su internet 

`{{ as_timestamp(states('sensor.date')) | timestamp_custom('%w') | int -1}}`

perché sulla documentazione di HA non ne stato riportato qualche esempio come questo?

se vado su 

https://www.home-assistant.io/integrations/template/#trigger-based-template-sensors

`{{ ( ( as_timestamp(now()) - as_timestamp(strptime("06.07.2018", "%d.%m.%Y")) ) / 86400 ) | round(default=0) }}`


Scusami sono da cellulare

**Comment by Metus88** [comment]
Magari può interessarvi, io uso [questo](https://github.com/bruxy70/Garbage-Collection) installabile via HACS e mi ha risolto tutto. Gestisce la raccolta stagionale e ricorsiva con qualsiasi o quasi tipo di alternanza... ed eventuali eccezioni per festività o altro.

**Reply by snakuzzo** [reply]
Secondo me, non di certo per sminuire il lavoro fatto dallo sviluppatore, per questa esigenza (un sensor template) non c'è bisogno di alcun custom component.
